### ISSUE TITLE
Misleading filter:notarray error

### ISSUE THREAD

**Issue by Andyhasit** [angular/angular.js#12720]
I made a simple error by not placing the "track by" instruction at the end:

     <tr ng-repeat="item in myItems| orderBy : 'name' track by $index | filter: {isUnmapped: true} ">

I should have put it here:

     <tr ng-repeat="item in myItems| orderBy : 'name' | filter: {isUnmapped: true} track by $index">

But the error raised for this mistake is filter:notarray, which is misleading:

Expected array but received: 0

Suggestions:
  * Create a new error if track by is in the wrong place: trackbyisnotlast
  * Check why the notarray got wrongly flagged despite an object being passed (and the alleged value passed is 0)

In AngularJS v1.4.4

**Comment by Narretz** [comment]
In the wrong example, the track by statement is inside a filter statement, and looks like an argument to the orderBy filter. The expression parser assumes `'name' track by $index` is the argument of the orderBy filter. In that sense, the error raised is correct, because the filter returns something that is not an array. Basically, we have no reliable way of detecting that the track by at this position is not a filter argument. Also, there is actually an error when track by appears before any filters.
Imo, this is a wontfix.

**Comment by gkalpak** [comment]
Actually, I think the problem is slightly different. The parser recognizes `track by $index | filter: {...}` as `track by ($index | filter: {...})`, i.e. `$index | filter: {...}` is recognized as the expression of `track by`. 
Since `track by` can take any valid Angular expression and since filters can appear in Angular expressions, there is no way we can know that this is a bug. 

In this particular case, since `$index` (number) is passed as input to `filterFilter`, which expects an array, an error is thrown.

But it could be an intended behaviour, e.g.: `...track by item.id | myCustomIdFilter`
(Where `myCustomIdFilter` returns the `id` if it exists or a random unique number if it doesn't exists, i.e. the instance has not been saved to the DB.)

--
Bottom line, this does indeed work as expected (so nothing to fix).

**Comment by Narretz** [comment]
Yes @gkalpak your explanation seems valid. If my guess was true, there should have been a parse error - maybe. In any case, this is really working as expected.
